walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
+     * Then it goes through events and removes free blocks when they overlap with an event.
+     * It then merges the remaining consecutive free blocks so they're outputted as one
+     * longer continuous TimeRange and we can verify that they're at least the length of the
+     * requested duration. Note that the function first tries to include both optional and mandatory
+     * attendees, but if no time ranges are found, only mandatory attendees are considered (if any).
+     */
+
+    // Get request information on attendees and duration
+    Collection<String> allAttendees = new ArrayList<String>(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    Collection<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees());
+    long meetingDuration = request.getDuration();
+    
+    // First attempt to include both mandatory and optional attendees
+    ArrayList<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+    // Initialize all free time blocks. Size 30 refers to the length of each block in minutes
+    // (ie. blocks are 12:00am, 12:30am, 1:00am etc) since we assume all events begin on the half hour
+    // Note that 15, 10, 5, 2, 1 minutes will all work but the granularity is not necessary
+    populateFreeTimes(freeTimes, 30);
+
+    // Remove from freeTimes all the TimeRanges where an attendee has an event
+    removeBusyTimes(freeTimes, events, allAttendees, 30);
+
+    // Make the remaining free blocks continuous and verify they are of minimum duration
+    updateMeetingTimes(meetingTimes, freeTimes, meetingDuration);
+    
+    // Verify that there is at least one potential meeting time with all attendees
+    // If not, only consider mandatory attendees (if any)
+    if (meetingTimes.size() != 0) {
+      return meetingTimes;
+    }
+
+    // Consider only mandatory attendees
+    if (mandatoryAttendees.size() != 0) {
+      freeTimes.clear();
+      populateFreeTimes(freeTimes, 30);
+      removeBusyTimes(freeTimes, events, mandatoryAttendees, 30);
+      updateMeetingTimes(meetingTimes, freeTimes, meetingDuration);
+    }
+    return meetingTimes;
+  }
+
+  private void populateFreeTimes (ArrayList<TimeRange> freeTimes, int blockLength) {
+    /* Populates FreeTimes with 1 day of blocks of size blockLength
+     * (i.e. blocklength 15 would mean 12:00am, 12:15am, 12:30am, 12:45am, etc)
+     * Note that blockLength must be a factor of 60 minutes or there may be unexpected behaviour.
+     */
+
+    // blocks holds all possible minutes value in the hour (ex: 0, 15, 30 if blockLength=15)
+    ArrayList<Integer> blocks = new ArrayList<Integer>();
+    for (int i = 0; i < 60; i++) {
+      if (i % blockLength == 0){
+        blocks.add(i);
+      }
+    }
+
+    // Save all TimeRanges
+    for (int i = 0; i < 24; i++) {
+      for (int b : blocks){
```

Comment:
That's true! Previously I noted that this technique would only work for blocks that were a factor of 60, so 45 wouldn't work. Great idea of making it a lot cleaner and simpler with the loops and eliminating the additional array. I switched it to just being 5 minute blocks across everything because that also removes needing to pass the exact parameter to every single function